### PR TITLE
fixed issue #9106

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Allow AQL query optimizer to use sparse indexes in more cases, specifically when indexes could not be used for filtering and there finally was an `EnumerateCollectionNode` in the query execution plan followed by a `SortNode`. In this case, sparse indexes were not considered for enumeration in sorted order, because it was unclear to the optimizer if the result set would contain null values or not.

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)

#### Related Information

- [x] There is a GitHub Issue reported by a Community User: https://github.com/arangodb/arangodb/issues/9106

### Testing & Verification

This change is already covered by existing tests, such as *tests/js/server/aql/aql-optimizer-indexes-sort.js*.

This PR adds a few new queries to the existing test suites to verify the changes.

https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/4635/

### Documentation

- [x] Added a *Changelog Entry*

### CODE REVIEW

File: arangod/Aql/Condition.cpp [hunk 1]
```diff
@@ -612,6 +613,65 @@ std::vector<std::vector<arangodb::basics::AttributeName>> Condition::getConstAtt
   return result;
 }
 
+/// @brief get the attributes for a sub-condition that are not-null
+std::unordered_set<std::vector<arangodb::basics::AttributeName>> Condition::getNonNullAttributes(
+    Variable const* reference) const {
+  std::unordered_set<std::vector<arangodb::basics::AttributeName>> result;
```
Comment (graetzer) [inline arangod/Aql/Condition.cpp]: why not use the HashSet ?

File: arangod/Aql/Condition.cpp [hunk 2]
```diff
@@ -612,6 +613,65 @@ std::vector<std::vector<arangodb::basics::AttributeName>> Condition::getConstAtt
...
+
+  if (_root == nullptr) {
+    return result;
+  }
+
+  size_t n = _root->numMembers();
```
Comment (graetzer) [inline arangod/Aql/Condition.cpp]: should you assert here that this is an OR condition

File: arangod/Aql/SortCondition.cpp [hunk 3]
```diff
@@ -152,6 +150,17 @@ SortCondition::SortCondition(
 
 /// @brief destroy the sort condition
 SortCondition::~SortCondition() {}
+  
+bool SortCondition::onlyUsesNonNullSortAttributes(
+    std::vector<std::vector<arangodb::basics::AttributeName>> const& attributes) const {
+  for (auto const& it : attributes) {
```
Comment (graetzer) [inline arangod/Aql/SortCondition.cpp]:
```suggestion
return std::all_of(attributes.begin(), attributes.end(), [this](auto const& it) { return _nonNullAttributes.find(it) == _nonNullAttributes.end()})
```

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/4648/

**Comment by jsteemann** [comment]
Tests blue

**Comment by jsteemann** [comment]
@graetzer @dothebart : can one of you please review the PR and either approve or disapprove? Thanks!

File: arangod/Aql/OptimizerRules.cpp [hunk 4]
```diff
@@ -2944,11 +2945,90 @@ struct SortToIndexNode final : public WalkerWorker<ExecutionNode> {
...
+      }
+      if (check != nullptr &&
+          check->isAttributeAccessForVariable(result, false) && 
+          result.first == variable) {
+        nonNullAttributes.emplace(std::move(result.second));
+      }
+    } else if (node->type == NODE_TYPE_OPERATOR_BINARY_EQ) {
+      AstNode const* lhs = node->getMember(0);
+      AstNode const* rhs = node->getMember(1);
+      AstNode const* check = nullptr;
```
Comment (graetzer) [inline arangod/Aql/OptimizerRules.cpp]: you can probably move these definitions up ?

File: arangod/Aql/OptimizerRules.cpp [hunk 5]
```diff
@@ -2944,11 +2945,90 @@ struct SortToIndexNode final : public WalkerWorker<ExecutionNode> {
...
+  void getNonNullAttributes(AstNode const* node, 
+                            Variable const* variable, 
+                            std::vector<std::vector<arangodb::basics::AttributeName>>& constAttributes,
+                            arangodb::HashSet<std::vector<arangodb::basics::AttributeName>>& nonNullAttributes) const {
+    std::pair<Variable const*, std::vector<arangodb::basics::AttributeName>> result;
+
+    if (node->type == NODE_TYPE_OPERATOR_BINARY_NE || 
+        node->type == NODE_TYPE_OPERATOR_BINARY_GT || 
+        node->type == NODE_TYPE_OPERATOR_BINARY_LT) {
+      AstNode const* lhs = node->getMember(0);
```
Comment (graetzer) [inline arangod/Aql/OptimizerRules.cpp]: can you use getMemberUnchecked ?

File: arangod/Aql/OptimizerRules.cpp [hunk 6]
```diff
@@ -2944,11 +2945,90 @@ struct SortToIndexNode final : public WalkerWorker<ExecutionNode> {
...
+          check->isAttributeAccessForVariable(result, false) && 
+          result.first == variable) {
+        nonNullAttributes.emplace(std::move(result.second));
+      }
+    } else if (node->type == NODE_TYPE_OPERATOR_BINARY_EQ) {
+      AstNode const* lhs = node->getMember(0);
+      AstNode const* rhs = node->getMember(1);
+      AstNode const* check = nullptr;
+
+      if (lhs->isConstant() && rhs->type == NODE_TYPE_ATTRIBUTE_ACCESS) {
```
Comment (graetzer) [inline arangod/Aql/OptimizerRules.cpp]:
```suggestion
      if (!lhs->isConstant() && rhs->type == NODE_TYPE_ATTRIBUTE_ACCESS) {
```

File: arangod/Aql/OptimizerRules.cpp [hunk 7]
```diff
@@ -2944,11 +2945,90 @@ struct SortToIndexNode final : public WalkerWorker<ExecutionNode> {
...
+        // const == doc.value
+        check = rhs;
+      } else if (rhs->isConstant() && lhs->type == NODE_TYPE_ATTRIBUTE_ACCESS) {
```
Comment (graetzer) [inline arangod/Aql/OptimizerRules.cpp]:
```suggestion
      } else if (rhs->isConstant() && !rhs->isNull() && lhs->type == NODE_TYPE_ATTRIBUTE_ACCESS) {
```

File: arangod/Aql/OptimizerRules.cpp [hunk 8]
```diff
@@ -2944,11 +2945,90 @@ struct SortToIndexNode final : public WalkerWorker<ExecutionNode> {
...
+      if (check != nullptr &&
+          check->isAttributeAccessForVariable(result, false) && 
+          result.first == variable) {
+        // found a constant value
+        constAttributes.emplace_back(std::move(result.second));
+      }
+
+    } else if (node->type == NODE_TYPE_OPERATOR_BINARY_AND) {
+      // recurse into both sides
+      getNonNullAttributes(node->getMember(0), variable, constAttributes, nonNullAttributes);
```
Comment (graetzer) [inline arangod/Aql/OptimizerRules.cpp]:
```suggestion
      getNonNullAttributes(node->getMemberUnchecked(0), variable, constAttributes, nonNullAttributes);
```

File: arangod/Aql/OptimizerRules.cpp [hunk 9]
```diff
@@ -2944,11 +2945,90 @@ struct SortToIndexNode final : public WalkerWorker<ExecutionNode> {
...
+      getNonNullAttributes(node->getMember(1), variable, constAttributes, nonNullAttributes);
```
Comment (graetzer) [inline arangod/Aql/OptimizerRules.cpp]:
```suggestion
      getNonNullAttributes(node->getMemberUnchecked(1), variable, constAttributes, nonNullAttributes);
```

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/4659/

File: arangod/Aql/OptimizerRules.cpp (hunk 6 above)
Comment (graetzer) [reply arangod/Aql/OptimizerRules.cpp]: does isConstant imply it is not null ? Otherwise wouldn't it match `FILTER null == doc.value` ?

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/4663/

**Pull request merged**